### PR TITLE
feat: `epochs_to_process`

### PR DESCRIPTION
**Pull request by vgorkavenko**

### CODE REVIEW

File: src/modules/csm/state.py [hunk 1]
```diff
@@ -75,13 +94,50 @@ def status(self) -> None:
...
+            self.clear()
+            self.commit()
+
+        try:
+            self.validate(l_epoch, r_epoch)
+        except InvalidStateLeftBorder:
+            _discard()
+        except InvalidStateRightBorder:
+            if self.r_epoch < r_epoch:
+                _discard()
```
Comment (madlabman) [inline src/modules/csm/state.py:126]: It should be `self.r_epoch > r_epoch` instead.

File: src/modules/csm/state.py [hunk 2]
```diff
@@ -75,13 +94,50 @@ def status(self) -> None:
...
+            else:
+                logger.warning({"msg": "The last report was missed. Reuse the state with the new right border"})
+                # expand the state to the new right border
+                self.epochs_to_process.update(sequence(self.r_epoch - 1, r_epoch))
```
Comment (madlabman) [inline src/modules/csm/state.py:130]: `self.r_epoch` as well as `self.r_epoch - 1` could be already processed, so `self.r_epoch + 1`, I think.

File: src/modules/csm/state.py [hunk 3]
```diff
@@ -75,13 +94,50 @@ def status(self) -> None:
...
+
+        if self.r_epoch > 0 and self.r_epoch != r_epoch:
+            logger.warning({"msg": f"Invalid right state border: should be {r_epoch=} but {self.r_epoch=}"})
+            raise InvalidStateRightBorder()
+
+    def validate_and_adjust(self, l_epoch: EpochNumber, r_epoch: EpochNumber):
+
+        def _discard():
+            logger.warning({"msg": "Discarding invalidated state cache"})
+            self.clear()
```
Comment (madlabman) [inline src/modules/csm/state.py:117]: `assert is_empty` inside the `clear` method would be nice.

File: src/modules/csm/state.py [hunk 4]
```diff
@@ -26,9 +39,10 @@ def perf(self) -> float:
 @dataclass
 class State(UserDict[ValidatorIndex, AttestationsAggregate]):
     """Tracks processing state of CSM performance oracle frame"""
-
+    l_epoch: EpochNumber = field(default_factory=int)
+    r_epoch: EpochNumber = field(default_factory=int)
     data: dict[ValidatorIndex, AttestationsAggregate] = field(default_factory=dict)
-    processed_epochs: set[EpochNumber] = field(default_factory=set)
+    epochs_to_process: set[EpochNumber] = field(default_factory=set)
```
Comment (madlabman) [inline src/modules/csm/state.py:45]: The thing I'm concerned the most is that we have no metric to check the state as it was using `processed_epochs`.

File: src/modules/csm/state.py [hunk 5]
```diff
@@ -75,13 +94,50 @@ def status(self) -> None:
...
+    def validate(self, l_epoch: EpochNumber, r_epoch: EpochNumber) -> None:
+        if self.l_epoch > 0 and self.l_epoch != l_epoch:
+            logger.warning({"msg": f"Invalid left state border: should be {l_epoch=} but {self.l_epoch=}"})
+            raise InvalidStateLeftBorder()
+
+        if self.r_epoch > 0 and self.r_epoch != r_epoch:
+            logger.warning({"msg": f"Invalid right state border: should be {r_epoch=} but {self.r_epoch=}"})
+            raise InvalidStateRightBorder()
+
+    def validate_and_adjust(self, l_epoch: EpochNumber, r_epoch: EpochNumber):
```
Comment (madlabman) [inline src/modules/csm/state.py:113]: `migrate` or something similar?

**Review by madlabman — CHANGES_REQUESTED** [review]

File: src/modules/csm/state.py (hunk 5 above)
Comment (vgorkavenko) [reply src/modules/csm/state.py:101]: not sure. left as is for a while

**Pull request merged**